src/language/data-representation.rkt

```diff
@@ -1,13 +1,19 @@
 #lang typed/racket
-(require (for-syntax racket/syntax))
+(require (for-syntax racket/syntax "c-helpers.rkt"))
 (require "../language/forms.rkt")
 (provide (all-defined-out))
 
 (define-syntax (define-constants stx)
   (syntax-case stx ()
     [(_ (name val) ...)
-     (let ([fmt (lambda (x) (format-id stx "data:~a" x))])
+     (let ([fmt (lambda (x) (format-id x "data:~a" x))])
```

Comment:
The correct syntax to use as the context for an identifier that you want to introduce (in a _client's_ scope) is usually the syntax from which the new name is derived. This is because this syntax is usually provided by the client, and thus has the client's scope information. Here we were using the syntax of the caller of this macro, which didn't end up being the client but an intermediary. 

Here I am using client (ie specify-representation.rkt) and intermediary (ie define-types.... macro) purposely vaguely because I don't have good terminology for what I would like to express. Hope it helps. If not I can try to express the idea better.

Reply:
That explains it!